server/src/main/java/org/apache/druid/metadata/BasicDataSourceExt.java

```diff
@@ -0,0 +1,179 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.metadata;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.commons.dbcp2.BasicDataSource;
+import org.apache.commons.dbcp2.ConnectionFactory;
+import org.apache.druid.java.util.common.RE;
+import org.apache.druid.java.util.common.logger.Logger;
+
+import java.sql.Driver;
+import java.sql.DriverManager;
+import java.sql.SQLException;
+import java.util.Properties;
+
+/**
+ * This class exists so that {@link MetadataStorageConnectorConfig} is asked for password every time a brand new
+ * connection is established with DB. {@link PasswordProvider} impls such as those based on AWS tokens refresh the
+ * underlying token periodically since each token is valid for a certain period of time only.
+ * So, This class overrides (and largely copies due to lack of extensibility), the methods from base class in order to keep
+ * track of connection properties and call {@link MetadataStorageConnectorConfig#getPassword()} everytime a new
+ * connection is setup.
+ */
+public class BasicDataSourceExt extends BasicDataSource
+{
+  private static final Logger LOGGER = new Logger(BasicDataSourceExt.class);
+
+  private Properties connectionProperties;
+  private final MetadataStorageConnectorConfig connectorConfig;
+
+  public BasicDataSourceExt(MetadataStorageConnectorConfig connectorConfig)
+  {
+    this.connectorConfig = connectorConfig;
+    this.connectionProperties = new Properties();
+  }
+
+  @Override
+  public void addConnectionProperty(String name, String value)
+  {
+    connectionProperties.put(name, value);
+    super.addConnectionProperty(name, value);
+  }
+
+  @Override
+  public void removeConnectionProperty(String name)
+  {
+    connectionProperties.remove(name);
+    super.removeConnectionProperty(name);
+  }
+
+  @Override
+  public void setConnectionProperties(String connectionProperties)
+  {
+    if (connectionProperties == null) {
+      throw new NullPointerException("connectionProperties is null");
+    }
+
+    String[] entries = connectionProperties.split(";");
+    Properties properties = new Properties();
+    for (String entry : entries) {
+      if (entry.length() > 0) {
+        int index = entry.indexOf('=');
+        if (index > 0) {
+          String name = entry.substring(0, index);
+          String value = entry.substring(index + 1);
+          properties.setProperty(name, value);
+        } else {
+          // no value is empty string which is how java.util.Properties works
+          properties.setProperty(entry, "");
+        }
+      }
+    }
+    this.connectionProperties = properties;
+    super.setConnectionProperties(connectionProperties);
+  }
+
+  @VisibleForTesting
+  public Properties getConnectionProperties()
+  {
+    return connectionProperties;
+  }
+
+  @Override
+  protected ConnectionFactory createConnectionFactory() throws SQLException
```

Comment:
@himanshug Did you adopt this source code from `org.apache.commons.dbcp2.BasicDataSource`? If so, it should be added in the licenses.yaml file.

Reply:
yeah, most code in this file is copied from `org.apache.commons.dbcp2.BasicDataSource` , will send a PR shortly.

Reply:
Thank you!